From 80061becfb10e17db05dc40914bee18b810ac70b Mon Sep 17 00:00:00 2001
From: Kai Sommerfeld <kai.sommerfeld@gmx.com>
Date: Wed, 21 Aug 2019 12:44:53 +0200
Subject: [PATCH] [listproviders] Update addons list on repository updated (for
 example, to instantly show 'updates available' on Estuary addons home screen
 widget).

---
 xbmc/listproviders/DirectoryProvider.cpp | 11 +++++++++++
 xbmc/listproviders/DirectoryProvider.h   |  2 ++
 2 files changed, 13 insertions(+)

diff --git a/xbmc/listproviders/DirectoryProvider.cpp b/xbmc/listproviders/DirectoryProvider.cpp
index d280adfa7d40..2de85b9c7505 100644
--- a/xbmc/listproviders/DirectoryProvider.cpp
+++ b/xbmc/listproviders/DirectoryProvider.cpp
@@ -294,6 +294,15 @@ void CDirectoryProvider::OnAddonEvent(const ADDON::AddonEvent& event)
   }
 }
 
+void CDirectoryProvider::OnAddonRepositoryEvent(const ADDON::CRepositoryUpdater::RepositoryUpdated& event)
+{
+  CSingleLock lock(m_section);
+  if (URIUtils::IsProtocol(m_currentUrl, "addons"))
+  {
+    m_updateState = INVALIDATED;
+  }
+}
+
 void CDirectoryProvider::OnPVRManagerEvent(const PVR::PVREvent& event)
 {
   CSingleLock lock(m_section);
@@ -337,6 +346,7 @@ void CDirectoryProvider::Reset()
     m_isAnnounced = false;
     CServiceBroker::GetAnnouncementManager()->RemoveAnnouncer(this);
     CServiceBroker::GetFavouritesService().Events().Unsubscribe(this);
+    CServiceBroker::GetRepositoryUpdater().Events().Unsubscribe(this);
     CServiceBroker::GetAddonMgr().Events().Unsubscribe(this);
     CServiceBroker::GetPVRManager().Events().Unsubscribe(this);
   }
@@ -461,6 +471,7 @@ bool CDirectoryProvider::UpdateURL()
     m_isAnnounced = true;
     CServiceBroker::GetAnnouncementManager()->AddAnnouncer(this);
     CServiceBroker::GetAddonMgr().Events().Subscribe(this, &CDirectoryProvider::OnAddonEvent);
+    CServiceBroker::GetRepositoryUpdater().Events().Subscribe(this, &CDirectoryProvider::OnAddonRepositoryEvent);
     CServiceBroker::GetPVRManager().Events().Subscribe(this, &CDirectoryProvider::OnPVRManagerEvent);
     CServiceBroker::GetFavouritesService().Events().Subscribe(this, &CDirectoryProvider::OnFavouritesEvent);
   }
diff --git a/xbmc/listproviders/DirectoryProvider.h b/xbmc/listproviders/DirectoryProvider.h
index 962984c93a5e..f47afc7c720c 100644
--- a/xbmc/listproviders/DirectoryProvider.h
+++ b/xbmc/listproviders/DirectoryProvider.h
@@ -10,6 +10,7 @@
 
 #include "IListProvider.h"
 #include "addons/AddonEvents.h"
+#include "addons/RepositoryUpdater.h"
 #include "favourites/FavouritesService.h"
 #include "guilib/GUIStaticItem.h"
 #include "interfaces/IAnnouncer.h"
@@ -84,6 +85,7 @@ class CDirectoryProvider :
   bool UpdateLimit();
   bool UpdateSort();
   void OnAddonEvent(const ADDON::AddonEvent& event);
+  void OnAddonRepositoryEvent(const ADDON::CRepositoryUpdater::RepositoryUpdated& event);
   void OnPVRManagerEvent(const PVR::PVREvent& event);
   void OnFavouritesEvent(const CFavouritesService::FavouritesUpdated& event);
   std::string GetTarget(const CFileItem& item) const;
